### PR TITLE
Added some docstrings and simplified parsing process

### PR DESCRIPTION
**Pull request by OrBin**
Adding docstrings is related to #32, but this PR does not close the issue.

### CODE REVIEW

File: gramhopper/configuration/trigger_response_parser.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+class TriggerResponseParser(abc.ABC):
+    """ A common base parser for trigger/response configuration parsers"""
+
+    def __init__(self, mapping_class: Type[DictEnum], element_base_class: Type[TriggerResponse]):
+        self.__mapping_class = mapping_class
+        self.__element_base_class = element_base_class
+
+    @property
+    def mapping_class(self) -> Type[DictEnum]:
```
Comment (meirhalachmi) [inline gramhopper/configuration/trigger_response_parser.py]: That's the pythonic way to make it read-only?

File: gramhopper/configuration/trigger_response_parser.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        # Some triggers (most of them) are classes and some are instances (mostly filter triggers).
+        # This allows both cases to be used.
+        if isclass(element):
+            trigger_or_response = element(**config_copy)
+            if name:
+                trigger_or_response.name = name
+            return trigger_or_response
+        return element
+
+    def parse_single(self, config: Union[CommentedMap, str], global_elements: GlobalsDict) \
```
Comment (meirhalachmi) [inline gramhopper/configuration/trigger_response_parser.py]: Naming again.
Single vs Atomic is confusing.
Maybe rename parse_single to parse_tree, and parse_all to parse_all_trees?

File: gramhopper/configuration/trigger_response_parser.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        Recursively parses a single configuration subtree as a trigger/response
+        :param config: A configuration subtree to parse
+        :param global_elements: A dictionary with all triggers and responses configured globally
+        :return: The trigger/response object created from the given configuration subtree
+        """
+        if isinstance(config, str):
+            return BooleanHelper.parse_boolean_subrule(config, global_elements)
+
+        config_copy = dict(config)
```
Comment (meirhalachmi) [inline gramhopper/configuration/trigger_response_parser.py]: What kind of copy?

File: gramhopper/configuration/common_types.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
-from typing import Union
+from typing import Union, Dict
 from ..triggers.basic_triggers import BaseTrigger
 from ..responses.basic_responses import BaseResponse
 
 
 TriggerResponse = Union[BaseTrigger, BaseResponse]
+GlobalsDict = Dict[str, TriggerResponse]
```
[the remark this answers is not in the document]
Comment (OrBin) [reply gramhopper/configuration/common_types.py]: The type is a dict of triggers and responses, which is used only for global ones, so I'm not sure it is such a bad name.
Also, it's not an excuse, but this is not a new type - it was just moved from another file.

File: gramhopper/configuration/common_types.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
-from typing import Union
+from typing import Union, Dict
 from ..triggers.basic_triggers import BaseTrigger
 from ..responses.basic_responses import BaseResponse
 
 
 TriggerResponse = Union[BaseTrigger, BaseResponse]
```
[the remark this answers is not in the document]
Comment (OrBin) [reply gramhopper/configuration/common_types.py]: We talked and decided to rename to "TriggerOrResponse" for now.

File: gramhopper/configuration/trigger_response_parser.py (hunk 1 above)
Comment (OrBin) [reply gramhopper/configuration/trigger_response_parser.py]: kind-of.
It's just a property with a getter and without a setter.

File: gramhopper/configuration/trigger_response_parser.py (hunk 2 above)
Comment (OrBin) [reply gramhopper/configuration/trigger_response_parser.py]: "Tree" sounds to me like it's all triggers (or all responses).

File: gramhopper/configuration/trigger_response_parser.py (hunk 3 above)
Comment (OrBin) [reply gramhopper/configuration/trigger_response_parser.py]: What do you mean?

File: gramhopper/configuration/trigger_response_parser.py (hunk 2 above)
Comment (OrBin) [reply gramhopper/configuration/trigger_response_parser.py]: Also, have a look at the naming in rule_parser.py.

File: gramhopper/configuration/trigger_response_parser.py (hunk 3 above)
Comment (meirhalachmi) [reply gramhopper/configuration/trigger_response_parser.py]: deep or shallow? 
what is the original type of `config` , also a dict?

Comment (OrBin) [reply gramhopper/configuration/trigger_response_parser.py]: Shallow.
The original type is `CommentedMap`, which inherits `dict`.

**Pull request merged**